internal/cmd/gocdk/launch.go

```diff
@@ -15,11 +15,170 @@
 package main
 
 import (
+	"bytes"
 	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/url"
+	"os/exec"
+	"path/filepath"
 
 	"golang.org/x/xerrors"
 )
 
 func launch(ctx context.Context, pctx *processContext, args []string) error {
-	return xerrors.New("not implemented")
+	f := newFlagSet(pctx, "launch")
+	imageName := f.String("image", "", "Docker image name to launch (defaults to project latest)")
+	if err := f.Parse(args); xerrors.Is(err, flag.ErrHelp) {
+		return nil
+	} else if err != nil {
+		return usagef("gocdk launch: %w", err)
+	}
+	if f.NArg() != 1 {
+		return usagef("gocdk launch BIOME")
+	}
+	biome := f.Arg(0)
+
+	moduleRoot, err := findModuleRoot(ctx, pctx.workdir)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Get the image name from the Dockerfile if not specified.
+	if *imageName == "" {
+		var err error
+		*imageName, err = moduleDockerImageName(moduleRoot)
+		if err != nil {
+			return xerrors.Errorf("gocdk launch: %w", err)
+		}
+	}
+
+	// Prepare the launcher.
+	cfg, err := readBiomeConfig(moduleRoot, biome)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+	if cfg.Launcher == nil {
+		return xerrors.Errorf("gocdk launch: launcher not specified in %s", filepath.Join(findBiomeDir(moduleRoot, biome), biomeConfigFileName))
+	}
+	launcher, err := newLauncher(ctx, pctx, *cfg.Launcher)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Read the launch specifier from the biome's Terraform output.
+	tfOutput, err := tfReadOutput(ctx,
+		findBiomeDir(moduleRoot, biome),
+		pctx.overrideEnv("TF_IN_AUTOMATION=1"))
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Launch the application.
+	launchURL, err := launcher.Launch(ctx, &launchInput{
```

Comment:
How are you choosing which arguments to pass to `Launch` vs which ones are passed to `newLauncher`? It might be simpler to put them all in once place or the other...?

Reply:
That which is specific to a launcher type and could be reused between launches goes into the launcher constructor. \
That which is generic to any launcher goes into the `launchInput`.

Reply:
But isn't `newLauncher` used for all launchers as well? I think your comment would make sense to me if there were a separate function being called to construct each launcher, but there's only one (with a switch inside it). So the args to `newLauncher` aren't really specific to a launcher type are they?

Reply:
It is broadly speaking helpful for the interface split, especially if we move forward with #2046 as we discussed offline.

Reply:
I'm fine with delaying this until then. Can you add an explicit TODO about figuring out which arguments should be passed to create the launcher, vs which arguments should be passed to `launcher.Launch`? It's not obvious, especially since `Launch` is likely only going to be called once.